include/benchmark/benchmark.h

```diff
@@ -1302,6 +1303,77 @@ struct SystemInfo {
   BENCHMARK_DISALLOW_COPY_AND_ASSIGN(SystemInfo);
 };
 
+// BenchmarkName wraps a std::string containing different fields
+// of a benchmark's name (e.g. the 'root' name, the benchmark's
+// 'args' etc.).
+class BenchmarkName {
+ public:
+  // Different fields of the benchmark's name.
+  enum Field {
+    // The name of the function being invoked in the benchmark.
+    kRoot = 1U,
+    // The arguments supplied to the benchmark, e.g. via Range().
+    kArgs = 1U << 1U,
+    // The minimum time specified for a benchmark.
+    kMinTime = 1U << 2U,
+    // The number of iterations specified for a benchmark.
+    kIterations = 1U << 3U,
+    // The number of repetitions specified for a benchmark.
+    kRepetitions = 1U << 4U,
+    // The type of time to measure for a benchmark.
+    kTimeType = 1U << 5U,
+    // The number of concurrent threads for a benchmark.
+    kThreads = 1U << 6U,
+    // All fields.
+    kAll =
+        kRoot | kArgs | kMinTime | kIterations | kRepetitions | kTimeType | kThreads
+  };
+
+  // Construct BenchmarkName with an 'empty' name.
+  BenchmarkName();
+
+  // Construct BenchmarkName with a given 'root'.
+  BenchmarkName(std::string root);
+
+  // Append 'field_content' to a given field of the name.
+  // Each appended element of content for a given field will be
+  // separated by a '/', and additionally each field will
+  // be separated by a '/'.
+  BenchmarkName& append(Field field, const std::string& field_content);
+
+  // Get the benchmark's name including all the fields set in 'fields'
+  std::string get(Field fields) const;
```

Comment:
we shouldn't need `str` below if we have this, as `str() == get(Field::kAll)`, right?

Reply:
True, although `get(Field::kAll)` would, currently, re-build the string component-wise whereas `.str()` just does a string copy. Could add a fast-path into the implementation of `get` though.